### ISSUE TITLE
OpenResty 1.15.8.3 breaks ngx.req.set_uri ability to set and proxy URLs with spaces (and other characters)

### ISSUE THREAD

**Issue by GUI** [openresty/lua-nginx-module#1676]
I understand the changes made in https://github.com/openresty/lua-nginx-module/pull/1654 were to address a security issue with `ngx.req.set_uri`. However, this seems to break various use-cases for `set_uri` that were working before, including for proxying seemingly safe characters, like spaces. So I was wondering if these changes for things like spaces were intentional and necessary, or is this a new bug?

With the changes in 1.15.8.3, you can't use `ngx.req.set_uri` with any special characters. However, attempting to URL-escape those special characters before passing them into `ngx.req.set_uri` then leads to the characters being double escaped when used with `proxy_pass`. So this leads to a situation where it seems like you can't really use `ngx.req.set_uri` with any input containing special characters (even when those special characters are valid or escaped).

As a quick example, if you were using something like `ngx.req.set_uri(ngx.var.uri)` along with a `proxy_pass` rule, this would work in 1.15.8.2 when proxying a URL like `/foo%20bar.pdf`. However, in 1.15.8.3 it seems like you're unable to utilize `set_uri` to set anything equivalent to that original URL. Calling `ngx.req.set_uri("/foo bar.pdf")` will result in the new "attempt to use unsafe uri" error being thrown. But if you try `ngx.req.set_uri("/foo%20bar.pdf")` that ends up getting proxied as `/foo%2520bar.pdf`, which is not equivalent.

It seems like this issue has come up before in slightly different contexts (https://github.com/openresty/lua-nginx-module/issues/659, https://github.com/openresty/lua-nginx-module/issues/1069), but in light of the new 1.15.8.3 release, I thought I'd check to see if the new behavior is in fact intentional in all these cases. If this new behavior is intentional, based on #1069, is this really an issue with nginx's `proxy_pass`? If so, is the suggested workaround simply to not use `ngx.req.set_uri` and rely on setting variables for use in `proxy_pass` (eg, `proxy_pass http://127.0.0.1:8080$new_uri`)?

### Example

Here's a set of servers and to demonstrate this issue:

```
server {
  listen 8080;
  location / {
    default_type text/plain;
    content_by_lua_block {
      ngx.say("ngx.var.uri: " .. ngx.var.uri)
      ngx.say("ngx.var.request_uri: " .. ngx.var.request_uri)
    }
  }
}

server {
  listen 8081;
  location / {
    proxy_pass http://127.0.0.1:8080/;
  }
}

server {
  listen 8082;
  location / {
    rewrite_by_lua_block {
      ngx.log(ngx.ERR, "ngx.var.uri: " .. ngx.var.uri)
      ngx.req.set_uri(ngx.var.uri)
    }

    proxy_pass http://127.0.0.1:8080/;
  }
}

server {
  listen 8083;
  location / {
    rewrite_by_lua_block {
      ngx.log(ngx.ERR, "ngx.var.request_uri: " .. ngx.var.request_uri)
      ngx.req.set_uri(ngx.var.request_uri)
    }

    proxy_pass http://127.0.0.1:8080/;
  }
}

server {
  listen 8084;
  location / {
    rewrite_by_lua_block {
      ngx.log(ngx.ERR, "ngx.escape_uri(ngx.var.uri): " .. ngx.escape_uri(ngx.var.uri))
      ngx.req.set_uri(ngx.escape_uri(ngx.var.uri))
    }

    proxy_pass http://127.0.0.1:8080/;
  }
}
```

Basically, under OpenResty 1.15.8.2, http://127.0.0.1:8082/foo%20bar.pdf returns:

```
ngx.var.uri: /foo bar.pdf
ngx.var.request_uri: /foo%20bar.pdf
```

Under OpenResty 1.15.8.3, http://127.0.0.1:8082/foo%20bar.pdf returns a 500 internal server error due to the `attempt to use unsafe uri` error being thrown.

Under OpenResty 1.15.8.3, attempts to use the escaped URL lead to the upstream server receiving a double-encoded URL, which isn't the same URL. http://127.0.0.1:8083/foo%20bar.pdf returns:

```
ngx.var.uri: /foo%20bar.pdf
ngx.var.request_uri: /foo%2520bar.pdf
```

**Comment by doujiang24** [comment]
@GUI Thanks for your report, we will take care of it soon :)

**Comment by doujiang24** [comment]
@GUI I think we can only mark the control characters as unsafe, ascii <= 31 and ascii == 127, Unicode charset are allowed then. what do you think about it?

**Comment by GUI** [comment]
@doujiang24: Yes, I think that sounds like a good solution. Thanks for looking into this so quickly! I'd assume this solution would still address https://hackerone.com/reports/513236, since control characters would be filtered, right? If I'm understanding that issue correctly, by filtering CR & LF characters, this vulnerability should still be addressed.

**Comment by doujiang24** [comment]
> If I'm understanding that issue correctly, by filtering CR & LF characters, this vulnerability should still be addressed.

@GUI In my opinion, by filtering `\0` is good enough only for that issue, but I think we'd better mark all control characters unsafe.
I have been making good progress on the new fix.

**Comment by thibaultcha** [comment]
This should be fixed in beb5c6b, thank you both!

**Comment by thibaultcha** [comment]
(And 1.15.8.4 is around the corder btw)

**Comment by jongiddy** [comment]
We've just hit this problem. Is there a schedule for the release of 1.15.8.4?

**Comment by sunnyakaxd** [comment]
Any schedule of this fix being ported to latest 1.17.8.x branch ??

**Comment by doujiang24** [comment]
it should be already fixed in OpenResty 1.17.8.* release.
http://openresty.org/en/ann-1017008002.html

**Comment by jongiddy** [comment]
I can confirm. I just tested the release versions 1.17.8.1 and .2, and they work.
